File: app/src/main/java/de/kuschku/quasseldroid_ng/ui/MessageAdapter.java
package de.kuschku.quasseldroid_ng.ui;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import de.kuschku.libquassel.Client;
import de.kuschku.libquassel.message.Message;
import de.kuschku.quasseldroid_ng.R;
import de.kuschku.util.observablelists.AutoScroller;
import de.kuschku.util.observablelists.ObservableSortedList;
import de.kuschku.util.observablelists.RecyclerViewAdapterCallback;

public class MessageAdapter extends RecyclerView.Adapter<MessageViewHolder> {
    private final AutoScroller scroller;
    private ObservableSortedList<Message> messageList = new ObservableSortedList<>(Message.class);
    private ChatMessageRenderer renderer;
    private LayoutInflater inflater;

    public MessageAdapter(Context ctx, AutoScroller scroller) {
        this.scroller = scroller;
        this.inflater = LayoutInflater.from(ctx);
        this.renderer = new ChatMessageRenderer(ctx);
    }

    public void setClient(Client client) {
        renderer.setClient(client);
    }

    public void setMessageList(ObservableSortedList<Message> messageList) {
        this.messageList.setCallback(null);
        this.messageList = messageList;
        this.messageList.setCallback(new RecyclerViewAdapterCallback(this, scroller));
        notifyDataSetChanged();
    }

    @Override
    public MessageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new MessageViewHolder(inflater.inflate(R.layout.widget_chatmessage, parent, false));
    }

    // TODO: REWRITE THIS
    @Override
    public void onBindViewHolder(MessageViewHolder holder, int position) {
        Message msg = messageList.list.get(position);
        renderer.onBind(holder, msg);
    }

    @Override
    public int getItemCount() {
        return messageList.list.size();
    }
}

File: app/src/main/java/de/kuschku/quasseldroid_ng/ui/chat/MessageAdapter.kt
package de.kuschku.quasseldroid_ng.ui.chat

import android.arch.lifecycle.LiveData
import android.arch.paging.PagedListAdapter
import android.content.Context
import android.support.v7.recyclerview.extensions.DiffCallback
import android.util.LruCache
import android.view.LayoutInflater
import android.view.ViewGroup
import de.kuschku.libquassel.protocol.*
import de.kuschku.libquassel.util.hasFlag
import de.kuschku.quasseldroid_ng.persistence.QuasselDatabase
import de.kuschku.quasseldroid_ng.persistence.QuasselDatabase.DatabaseMessage
import de.kuschku.quasseldroid_ng.ui.settings.data.RenderingSettings
import de.kuschku.quasseldroid_ng.util.helper.getOrPut

class MessageAdapter(
  context: Context,
  private val markerLine: LiveData<Pair<MsgId, MsgId>?>
) : PagedListAdapter<QuasselDatabase.DatabaseMessage, QuasselMessageViewHolder>(
  object : DiffCallback<QuasselDatabase.DatabaseMessage>() {
    override fun areItemsTheSame(oldItem: QuasselDatabase.DatabaseMessage,
                                 newItem: QuasselDatabase.DatabaseMessage)
      = DatabaseMessage.MessageDiffCallback.areItemsTheSame(oldItem, newItem)

    override fun areContentsTheSame(oldItem: QuasselDatabase.DatabaseMessage,
                                    newItem: QuasselDatabase.DatabaseMessage)
      = DatabaseMessage.MessageDiffCallback.areContentsTheSame(oldItem, newItem) &&
        oldItem.messageId != markerLine.value?.first &&
        oldItem.messageId != markerLine.value?.second
  }
) {
  private val messageRenderer: MessageRenderer = QuasselMessageRenderer(
    context,
    RenderingSettings(
      showPrefix = RenderingSettings.ShowPrefixMode.FIRST,
      colorizeNicknames = RenderingSettings.ColorizeNicknamesMode.ALL_BUT_MINE,
      colorizeMirc = true,
      timeFormat = ""
    )
  )

  private val messageCache = LruCache<Int, FormattedMessage>(512)

  override fun onBindViewHolder(holder: QuasselMessageViewHolder, position: Int) {
    getItem(position)?.let {
      messageRenderer.bind(
        holder,
        if (it.messageId == markerLine.value?.second || it.messageId == markerLine.value?.first) {
          val value = messageRenderer.render(it, markerLine.value?.second ?: -1)
          messageCache.put(it.messageId, value)
          value
        } else {
          messageCache.getOrPut(it.messageId) {
            messageRenderer.render(it, markerLine.value?.second ?: -1)
          }
        }
      )
    }
  }

  override fun getItemViewType(position: Int): Int {
    val item = getItem(position)
    if (item != null) {
      return viewType(Message_Flags.of(item.type), Message_Flags.of(item.flag))
    } else {
      return 0
    }
  }

  private fun viewType(type: Message_Types, flags: Message_Flags): Int {
    if (flags.hasFlag(Message_Flag.Highlight)) {
      return -type.value
    } else {
      return type.value
    }
  }

  override fun getItemId(position: Int): Long {
    return getItem(position)?.messageId?.toLong() ?: 0L
  }

  private fun messageType(viewType: Int): Message_Type?
    = Message_Type.of(Math.abs(viewType)).enabledValues().firstOrNull()

  private fun hasHiglight(viewType: Int)
    = viewType < 0

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): QuasselMessageViewHolder {
    val messageType = messageType(viewType)
    val hasHighlight = hasHiglight(viewType)
    val viewHolder = QuasselMessageViewHolder(
      LayoutInflater.from(parent.context).inflate(
        messageRenderer.layout(messageType, hasHighlight),
        parent,
        false
      )
    )
    messageRenderer.init(viewHolder, messageType, hasHighlight)
    return viewHolder
  }

  operator fun get(position: Int): QuasselDatabase.DatabaseMessage? {
    return getItem(position)
  }
}

